text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+
+```rust
+match foo.branch() {
+    ControlFlow::Break(err) => {
+        yield R::from_residual(err);
+        return;
+    },
+    ControlFlow::Continue(val) => val,
+}
+```
+
+This is the same behaviour that `collect::<Result<_, _>>()` performs
+on iterators over `Result`s
+
+## Implementation
+
+This feature is mostly implemented via existing generators.
+We'll need additional desugarings and lots of work to get good diagnostics.
+
+### `#[rustc_gen] fn`
+
+`#[rustc_gen] fn` desugars to the function itself with the return type replaced by `impl Iterator<Item = $ret>` and its body wrapped in a `#[rustc_gen]` block.
+A `#[rustc_gen] fn`'s "return type" is its iterator's `yield` type.
+
+A `#[rustc_gen] fn` captures all lifetimes and #[rustc_gen]eric parameters into the `impl Iterator` return type (just like `async fn`).
+If more control over captures is needed, type alias impl trait can be used when it is stabilized.
+
+Like other uses of `impl Trait`, auto traits are revealed without being specified.
+
+### `#[rustc_gen]` blocks
+
+`#[rustc_gen]` blocks are the same as an unstable generator
+
+* without arguments,
+* with an additional check forbidding holding borrows across `yield` points,
+* and an automatic `Iterator` implementation.
+
+We'll probably be able to modularize the generator implementation and make it more robust on the implementation and diagnostics side for the `#[rustc_gen]` block case, but I believe the initial implementation should be a HIR lowering to a generator and wrapping that generator in [`from_generator`][].
+
+## Fusing
+
+Special case the generator implementation if `Generator::Return = ()` to not panic, but
+repeatedly produce values of the unit type.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's another language feature for something that can already be written entirely in user code.
+
+In contrast to `Generator`, `#[rustc_gen]` blocks that produce `Iterator`s cannot hold references across `yield` points.
+See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.
+
+[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+## Keyword
+
+We could use `iter` as the keyword.
+I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.
+The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.
+I believe connecting the `Iterator` trait with `iter` blocks is the right choice,
+but that would require us to carve out many exceptions for this keyword as `iter` is used for module names and method names everywhere (including libstd/libcore).
+It may not be much worse than `gen` (see also [the unresolved questions][#unresolved-questions]).
+We may want to use `gen` for full on generators in the future.
```

Comment:
If we were to use `gen` for full-on generators in the future, would they look something like `gen |arg1, arg2, ...| { ... }` ?

I.e. in that world, would the iterator-oriented feature in this RFC "continue" to be written `gen { ... }` (which I assume is the future expected state of `#[rustc_gen] { ... }`, and one would be distinguishing that from full-on generators soley via the presence/absence of the parameter list `|arg1, arg2, ...|` ?

(I'm not commenting yet on whether that mode of distinguishing is good or bad. I just want to make sure I understand all this first.)

Reply:
Yea, full on generators/coroutines need new syntax, but the `gen` keyword may also be reused in that syntax

Reply:
I'll try my best to answer this.

I don't think it can ever be enough to distinguish between the two kinds of generators just based on a parameter lists: after all this RFC already introduces "generator functions" with input parameters. Which means it's not unthinkable we'd want to enable "generator closures" to work in the future too.

For clarity I'll start referring to "full-on generators" as "co-routines" going forward instead. Co-routines are a lot more complicated than the current iterator is. They cannot be mapped to the existing `Iterator` trait, and instead require [a new trait](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) which can take arguments into the yield statement. As well as distinguish between a "yield type" and a "return type". Put together this would probably need to look something like this:

```rust
// example trait definition
pub trait CoRoutine<R = ()> {
    type Yield;
    type Return;
    fn resume(&mut Self, resume: R) -> CoRoutineState<Self::Yield, Self::Return>;
}
pub enum CoRoutineState <Y, R> {
    Yielded(Y),
    Complete(R),
}

// syntax example (entirely made up, not an actual proposal)
co(Food) fn eat_food(cat: &mut Cat) -> Nap yields Purr {
// ^                                   ^          ^
// |                                   |          |
// |                                   |  The type yielded (output)
// |                                   |
// |                           The return type (output)
// |
// The type returned from `yield` (input)
    for 0..10 {
        // We yield a `Purr`, but `yield` returns `Food`.
        let food = yield Purr::new();
        cat.eat(food);
    }
    Nap::new() // We end the function by returning a `Nap`
}
```

If you compare this with the `gen fn` feature proposed by this RFC, you'll notice that "co-routines" contain a lot more moving parts which need to be dfeined. And even if both features are conceptually related, they are actually quite different - and we should probably think of them as separate. Does that make sense?